### ISSUE TITLE
Introspection sinks don't survive compute reconciliation

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#14346]
### What version of Materialize are you using?

v0.26.1-dev (f029b1410)

### How did you install Materialize?

Built from source

### What is the issue?

Querying the new persist-based introspection sources hangs forever after a restart of `environmentd` (but not `computed`).

### Repro

Run `environmentd` + `computed` and then force an `environmentd` restart:

```bash
bin/environmentd --reset &
killall -9 environmentd
bin/environmentd
```

Then, connect via psql and try to query one of the persist-based introspection sources:

```sql
SELECT * from mz_materializations_1;
```

Observe that the query hangs forever. Also observe that `computed` produces the following logs indefinitely:

```
compute-cluster-1-replica-1: [...]  INFO mz_persist_client::internal::machine: snapshot [...] as of Antichain { elements: [1661240493686] } not yet available for upper Antichain { elements: [1661240424000] } retrying in 64ms
compute-cluster-1-replica-1: [...]  INFO mz_persist_client::internal::machine: snapshot [...] as of Antichain { elements: [1661240493686] } not yet available for upper Antichain { elements: [1661240424000] } retrying in 128ms
compute-cluster-1-replica-1: [...]  INFO mz_persist_client::internal::machine: snapshot [...] as of Antichain { elements: [1661240493686] } not yet available for upper Antichain { elements: [1661240424000] } retrying in 256ms
compute-cluster-1-replica-1: [...]  INFO mz_persist_client::internal::machine: snapshot [...] as of Antichain { elements: [1661240493686] } not yet available for upper Antichain { elements: [1661240424000] } retrying in 512ms
compute-cluster-1-replica-1: [...]  INFO mz_persist_client::internal::machine: snapshot [...] as of Antichain { elements: [1661240493686] } not yet available for upper Antichain { elements: [1661240424000] } retrying in 1.024s
...
```

### Analysis

The logs indicate that the introspection sinks never catch up with the current query time, which explains why the queries hang forever.

Debugging reveals that this is because no `FrontierUppers` responses are sent for introspection sinks after the `environmentd` restart. This in turn is the case because the respective entries in [ComputeState::sink_write_frontiers](https://github.com/MaterializeInc/materialize/blob/581271f8419b28ebabfdd60d0e4afcb077da644d/src/compute/src/compute_state.rs#L68) are not advanced anymore. That advancement should happen periodically [inside the persist_sink](https://github.com/MaterializeInc/materialize/blob/main/src/compute/src/sink/persist_sink.rs#L210) of each introspection dataflow. 

Further debugging reveals that the introspection persist_sinks after the `environmentd` restart discover that [their exit condition](https://github.com/MaterializeInc/materialize/blob/581271f8419b28ebabfdd60d0e4afcb077da644d/src/compute/src/sink/persist_sink.rs#L165-L170) is fulfilled, and therefore stop processing anymore inputs. The exit condition is fulfilled because the `token_weak.upgrade().is_none()` check becomes `true`.

So, upon `environmentd` reconnect, the tokens of the introspection sinks are dropped, which in turn makes the introspection sinks terminate themselves.

It looks like the token dropping happens during reconciliation [here](https://github.com/MaterializeInc/materialize/blob/581271f8419b28ebabfdd60d0e4afcb077da644d/src/compute/src/server.rs#L500-L502). My assumption is that the introspection sinks are not added to `retain_ids`, so this code isn't aware of the fact that they are still needed, so their tokens are dropped.

### Possible Solutions

If the above analysis is correct, fixing this should be simple. We can either:

1. Adapt the reconciliation code to add the IDs of introspection dataflows to `retain_ids`.
2. Adapt [the logging persist_sink code](https://github.com/MaterializeInc/materialize/blob/581271f8419b28ebabfdd60d0e4afcb077da644d/src/compute/src/logging/persist.rs#L45-L53) to not add the introspection sink tokens to `ComputeState::sink_tokens` in the first place, but instead leak them (`std::mem::forget`), to ensure the dataflows are never dropped.

**Comment by aljoscha** [comment]
@teskje Just making sure you also saw this one: https://github.com/MaterializeInc/materialize/issues/14261

**Comment by teskje** [comment]
Thanks! I actually discovered this while reviewing the PR that fixes #14261 (https://github.com/MaterializeInc/materialize/pull/14310).
